### PR TITLE
[SPARK-49474][SS] Classify Error class for FlatMapGroupsWithState user function error

### PR DESCRIPTION
**Pull request by liviazhu**
### What changes were proposed in this pull request?

Add new error classification for errors occurring in the user function that is used in FlatMapGroupsWithState. 

### Why are the changes needed?

The user provided function can throw any type of error. Using the new error framework for better error messages and classification.

### Does this PR introduce _any_ user-facing change?

Yes, better error message with error class for Foreach sink user function failures.

### How was this patch tested?

Updated existing tests and added new unit test in FlatMapGroupsWithStateSuite. 

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala [hunk 1]
```diff
@@ -635,6 +634,25 @@ class FlatMapGroupsWithStateSuite extends StateStoreMetricsTest {
     )
   }
 
+  testWithAllStateVersions("[SPARK-49474] flatMapGroupsWithState - user error is classified") {
+    // Function to maintain running count up to 2, and then remove the count
+    // Returns the data and the count if state is defined, otherwise does not return anything
+    val stateFunc = (_: String, _: Iterator[String], _: GroupState[RunningCount]) => {
+      Iterator(null.asInstanceOf[GroupState[RunningCount]].getCurrentProcessingTimeMs())
```
Comment (jerrypeng) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala:641]: An NPE will be thrown here?  If you are just trying to simulate an error in the user defined function, can we just explicitly throw an exception so the code is clear?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala [hunk 2]
```diff
@@ -447,8 +450,18 @@ case class FlatMapGroupsWithStateExec(
...
+        func(keyObj, valueObjIter, groupState)
+      } catch {
+        case NonFatal(e) if !e.isInstanceOf[SparkThrowable] =>
+          throw FlatMapGroupsWithStateUserFuncException(e)
+        case f: Throwable =>
+          throw f
+      }
+
+      // Get the returned objects and convert them to rows
+      val mappedIterator = iterator.map { obj =>
```
Comment (jerrypeng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala:464]: What if the user error occurs in the iterator?  That is, if a user returns lazy iterator with custom code.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala [hunk 3]
```diff
@@ -447,10 +450,17 @@ case class FlatMapGroupsWithStateExec(
         hasTimedOut,
         watermarkPresent)
 
-      // Call function, get the returned objects and convert them to rows
-      val mappedIterator = func(keyObj, valueObjIter, groupState).map { obj =>
-        numOutputRows += 1
-        getOutputRow(obj)
+      val mappedIterator = try {
+        // Call function, get the returned objects and convert them to rows
+        func(keyObj, valueObjIter, groupState).map { obj =>
```
Comment (jerrypeng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala:455]: I don't think this catches the issue I mentioned.  The code in the custom iterator will not get executed here.  For example, a user can return an iterator like the following

```
 new Iterator[Int] {

      override def hasNext: Boolean = {
        true
      }

      override def next(): Int = {
        throw new Exception()
      }

```

Only when "next" gets call will the problem occur.  I think you need to wrap the iterator returned by the user func and catch any exceptions there.

Can you also add a test for this scenario as well?

**Review by jerrypeng — APPROVED** [review]

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Failing tests are spark-connect side and known to be flaky.

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**